File: test/Zorglub.Tests.CSharp/Time/Core/UnitTests.cs
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2020 Narvalo.Org. All rights reserved.

namespace Zorglub.Time.Core;

using Zorglub.Time.Core.Utilities;

public static class UnitTests
{
    [Fact]
    public static void Constructor() =>
        Assert.Equal(default, new Unit());

    [Fact]
    public static void ToString_InvariantCulture() =>
        Assert.Equal("()", Unit.Value.ToString());

    [Fact]
    public static void DefaultValue() =>
        Assert.Equal(default, Unit.Value);

    [Fact]
    public static void Equals_WhenSame()
    {
        var unit = new Unit();
        var same = new Unit();

        // Act & Assert
        Assert.True(unit == same);
        Assert.False(unit != same);
        Assert.True(unit.Equals(same));
        Assert.True(unit.Equals((object)same));
    }

    [Fact]
    public static void Equals_ValueTuple()
    {
        var unit = new Unit();
        var tupl = new ValueTuple();
        // Act & Assert
        Assert.True(unit == tupl);
        Assert.True(tupl == unit);
        Assert.False(unit != tupl);
        Assert.False(tupl != unit);
        Assert.True(unit.Equals(tupl));
        Assert.True(unit.Equals((object)tupl));
    }

    [Fact]
    public static void Equals_NullOrOtherType()
    {
        var unit = new Unit();
        // Act & Assert
        Assert.False(unit.Equals(1));
        Assert.False(unit.Equals(null));
        Assert.False(unit.Equals(new object()));
    }

    [Fact]
    public static void HashCode() =>
        Assert.Equal(0, Unit.Value.GetHashCode());
}

File: test/Zorglub.Tests/Tests/Core/Utilities/UnitTests.fs
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) Tran Ngoc Bich. All rights reserved.

module Zorglub.Tests.Core.Utilities.UnitTests

open System

open Zorglub.Testing

open Zorglub.Time.Core.Utilities

open Xunit

module Prelude =
    [<Fact>]
    let ``Default value`` () =
        Unchecked.defaultof<Unit> === Unit.Value

    [<Fact>]
    let ``Constructor`` () =
        let unit = new Unit()

        unit === Unit.Value

    [<Fact>]
    let ``ToString()`` ()=
        Unit.Value.ToString() === "()"

module Equality =
    open NonStructuralComparison

    [<Fact>]
    let ``Equality when both operands are identical`` () =
        let x = Unit.Value

        x = x |> ok
        x <> x |> nok
        x.Equals(x) |> ok
        x.Equals(x :> obj) |> ok

    [<Fact>]
    let ``Equality with ValueTuple`` () =
        let x = Unit.Value
        let y = new ValueTuple()

        Unit.op_Equality(x, y) |> ok
        Unit.op_Equality(y, x) |> ok
        Unit.op_Inequality(x, y) |> nok
        Unit.op_Inequality(y, x) |> nok
        x.Equals(y) |> ok
        x.Equals(y :> obj) |> ok

    [<Fact>]
    let ``Equals(obj) returns false when "obj" is null or is a plain object`` () =
        let x = Unit.Value

        x.Equals(null) |> nok
        x.Equals(new obj()) |> nok

    [<Fact>]
    let ``GetHashCode() returns 0`` () =
        Unit.Value.GetHashCode() === 0
